### PR TITLE
fix for mutex slim being released too early during flush operation

### LINKED ISSUE #1438: Get & Set Failing After Error Cascade
Aloha!

### The Problem
Under certain loads and after long up-times, we've begun seeing a curious error during addorset (Reproduced at the end).
This is accompanied by a much higher number of errors for Gets (Reproduced at the end) that are roughly identical, though we do use get far more than set.

### Additional
Once a client becomes symptomatic, it rapidly begins throwing these errors on most or all operations. Bouncing that client causes another client to become symptomatic in a similar way, which may be a red herring but was interesting enough to bear mention.

We currently use only one multiplexer, which seems like a possible cause, but before we refactor, I thought I'd check in. Our application is pretty much pure C#. We are seeing time-outs as well, during symptomatic periods. My assumption is that [StackExchange.Redis/ConnectionMultiplexer.cs#L2601](https://github.com/StackExchange/StackExchange.Redis/blob/master/src/StackExchange.Redis/ConnectionMultiplexer.cs#L2601) is the top level call in the multiplexer for the error, given that 2622 is part of error handling deferral. 

### Set Error:
```
REDACTED REDIS AddOrSet Failure on key REDACTED
    StackExchange.Redis.RedisConnectionException: Failed to write ---> System.InvalidOperationException: Concurrent reads or writes are not supported.
       at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
       at System.IO.Pipelines.PipeCompletion.ThrowLatchedException()
       at System.IO.Pipelines.Pipe.GetFlushResult(FlushResult& result)
       at System.IO.Pipelines.Pipe.PrepareFlush(CompletionData& completionData, ValueTask`1& result, CancellationToken cancellationToken)
       at System.IO.Pipelines.Pipe.FlushAsync(CancellationToken cancellationToken)
       at System.IO.Pipelines.Pipe.DefaultPipeWriter.FlushAsync(CancellationToken cancellationToken)
       at Pipelines.Sockets.Unofficial.SocketConnection.WrappedWriter.FlushAsync(CancellationToken cancellationToken) in /_/src/Pipelines.Sockets.Unofficial/SocketConnection.cs:line 434
       at StackExchange.Redis.PhysicalConnection.FlushAsync(Boolean throwOnFailure) in /_/src/StackExchange.Redis/PhysicalConnection.cs:line 907
       at StackExchange.Redis.PhysicalBridge.ProcessBacklog() in /_/src/StackExchange.Redis/PhysicalBridge.cs:line 889
       --- End of inner exception stack trace ---
       at StackExchange.Redis.ConnectionMultiplexer.ExecuteSyncImpl[T](Message message, ResultProcessor`1 processor, ServerEndPoint server) in /_/src/StackExchange.Redis/ConnectionMultiplexer.cs:line 2622
       at StackExchange.Redis.RedisBase.ExecuteSync[T](Message message, ResultProcessor`1 processor, ServerEndPoint server) in /_/src/StackExchange.Redis/RedisBase.cs:line 54
```

### Get Error:
```
REDACTED REDIS GetAsync Failure on key REDACTED
    StackExchange.Redis.RedisConnectionException: Failed to write ---> System.InvalidOperationException: Concurrent reads or writes are not supported.
       at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
       at System.IO.Pipelines.PipeCompletion.ThrowLatchedException()
       at System.IO.Pipelines.Pipe.GetFlushResult(FlushResult& result)
       at System.IO.Pipelines.Pipe.PrepareFlush(CompletionData& completionData, ValueTask`1& result, CancellationToken cancellationToken)
       at System.IO.Pipelines.Pipe.FlushAsync(CancellationToken cancellationToken)
       at System.IO.Pipelines.Pipe.DefaultPipeWriter.FlushAsync(CancellationToken cancellationToken)
       at Pipelines.Sockets.Unofficial.SocketConnection.WrappedWriter.FlushAsync(CancellationToken cancellationToken) in //src/Pipelines.Sockets.Unofficial/SocketConnection.cs:line 434
       at StackExchange.Redis.PhysicalConnection.FlushAsync(Boolean throwOnFailure) in //src/StackExchange.Redis/PhysicalConnection.cs:line 907
       at StackExchange.Redis.PhysicalBridge.WriteMessageTakingWriteLockAsync(PhysicalConnection physical, Message message) in /_/src/StackExchange.Redis/PhysicalBridge.cs:line 987
       --- End of inner exception stack trace ---
       at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
       at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
```

### PR DESCRIPTION
**Pull request by arsnyder16**
Fixes #1438 

When flushing during FlushSync we are specifying a timeout to the Task, but when this timeout gets hit the underlying task is still running. The thread that invoked FlushSync continues and releases the MutexSlim that is protecting the io pipe allowing another thread to obtain the MutexSlim and try to run operations on the pipe. 

To fix use the cancellation token that can be passed to PipeWriter.FlushAsync. 

An alternative to this is to not specify a timeout. There are 2 additional locations that flush the pipe neither of those locations specify a timeout they just simply await the Task.    Generally the timeouts are driven by how long an operation waits to obtain the MutexSlim

### CODE REVIEW

**Comment by arsnyder16** [comment]
Fixes StackExchange/StackExchange.Redis#1438

**Comment by arsnyder16** [comment]
Looks like the CI checks are failing for some unrelated issue, seems like the other PRs have the same issue

**Comment by mgravell** [comment]
Normally, time constraints are limiting how much I can get done on PRs etc,
but this is very interesting. I'm going to carve out some of tomorrow
morning to look at this, because if you're right (and I'm assuming you are,
haven't looked deep at the code yet) it could be really useful! Thanks.

**Comment by arsnyder16** [comment]
Thanks, i would be interested in hearing your thoughts, since i am not overly familiar with the library. Feel free to adjust it in another way, this seemed to make the most sense to me.

File: src/StackExchange.Redis/PhysicalConnection.cs [hunk 1]
```diff
@@ -875,11 +875,23 @@ internal static int WriteRaw(Span<byte> span, long value, bool withLengthPrefix
         [Obsolete("this is an anti-pattern; work to reduce reliance on this is in progress")]
         internal WriteResult FlushSync(bool throwOnFailure, int millisecondsTimeout)
         {
-            var flush = FlushAsync(throwOnFailure);
+            CancellationTokenSource source = new CancellationTokenSource();
```
Comment (Plasma) [inline src/StackExchange.Redis/PhysicalConnection.cs:878]: We need to wrap this in a using/dispose block

**Comment by sassywarsat** [comment]
I can't speak to the exact fix, but the described behavior matches exactly the symptoms we've seen provoke the issue.

**Comment by sassywarsat** [comment]
Okay, I've walked through the precise fixes. Normally, I'd like to see a continuation bound for on cancel but that's not really the model or style of the lib, I think?

File: src/StackExchange.Redis/PhysicalConnection.cs (hunk 1 above)
Comment (arsnyder16) [reply src/StackExchange.Redis/PhysicalConnection.cs]: good catch. fixed.

**Comment by Plasma** [comment]
I agree the timeout can cause the lock to be released (incorrectly) while the task is actually making progress in the background. so I think this change is correct.

**Comment by NickCraver** [comment]
Merging main in here for updated build checks

**Comment by JKurzer** [comment]
Do we have a timeline for accepting the PR?

**Comment by mgravell** [comment]
Sorry, I had planned to already look at it, but things got ... busy. I hope to look tomorrow (although I appreciate that I've said that before)

**Comment by JKurzer** [comment]
No worries, we're in a stable-ish state on an older version, and life is pretty complicated for all of us right now.

**Comment by mgravell** [comment]
I concur with the problem, but I'm very dubious of allocating a CTS with scheduled timeout *every* call; I wonder instead if we could allocate and reuse a CTS, and only schedule the timeout (dooming it for reuse) when we get a non-sync response; something like this: https://github.com/StackExchange/StackExchange.Redis/commit/e1e7189d452bcf2527defa9c6714459d02bdcfc5

thoughts?

**Comment by mgravell** [comment]
merged with the CTS tweak as proposed

**Pull request merged**